internal/semantic/version-redhat.go

```diff
@@ -0,0 +1,235 @@
+func isASCIIDigit(c rune) bool {
+	return c >= 48 && c <= 57
+}
+
+// isASCIILetter returns true if the given rune is an ASCII letter.
+//
+// Unicode letters are not considered ASCII letters by this function.
+func isASCIILetter(c rune) bool {
+	return (c >= 65 && c <= 90) || (c >= 97 && c <= 122)
+}
+
+// shouldBeTrimmed checks if the given rune should be trimmed when parsing RedHatVersion components
+func shouldBeTrimmed(r rune) bool {
+	return !isASCIILetter(r) && !isASCIIDigit(r) && r != '~' && r != '^'
+}
+
+// compareRedHatComponents compares two components of a RedHatVersion in the same
+// manner as rpmvercmp(8) does.
+func compareRedHatComponents(a, b string) int {
+	if a == "" && b != "" {
+		return -1
+	}
+	if a != "" && b == "" {
+		return +1
+	}
+
+	var vi, wi int
+
+	for {
+		// 1. Trim anything that’s not [A-Za-z0-9] or tilde (~) from the front of both strings.
+		for {
+			if vi == len(a) || !shouldBeTrimmed(rune(a[vi])) {
+				break
+			}
+			vi++
+		}
+
+		for {
+			if wi == len(b) || !shouldBeTrimmed(rune(b[wi])) {
+				break
+			}
+			wi++
+		}
+
+		// 2. If both strings start with a tilde, discard it and move on to the next character.
+		vStartsWithTilde := vi < len(a) && a[vi] == '~'
+		wStartsWithTilde := wi < len(b) && b[wi] == '~'
+
+		if vStartsWithTilde && wStartsWithTilde {
+			vi++
+			wi++
+
+			continue
+		}
+
+		// 3. If string `a` starts with a tilde and string `b` does not, return -1 (string `a` is older); and the inverse if string `b` starts with a tilde and string `a` does not.
+		if vStartsWithTilde {
+			return -1
+		}
+		if wStartsWithTilde {
+			return +1
+		}
+
+		// 4. If both strings start with a caret, discard it and move on to the next character.
+		vStartsWithCaret := vi < len(a) && a[vi] == '^'
+		wStartsWithCaret := wi < len(b) && b[wi] == '^'
+
+		if vStartsWithCaret && wStartsWithCaret {
+			vi++
+			wi++
+
+			continue
+		}
+
+		// 5. if string `a` starts with a caret and string `b` does not, return -1 (string `a` is older) unless string `b` has reached zero length, in which case return +1 (string `a` is newer); and the inverse if string `b` starts with a caret and string `a` does not.
+		if vStartsWithCaret {
+			if wi == len(b) {
+				return +1
+			}
+
+			return -1
+		}
+		if wStartsWithCaret {
+			if vi == len(a) {
+				return -1
+			}
+
+			return +1
+		}
+
+		// 6. End the loop if either string has reached zero length.
+		if vi == len(a) || wi == len(b) {
+			break
+		}
+
+		// 7. If the first character of `a` is a digit, pop the leading chunk of continuous digits from each string (which may be "" for `b` if only one `a` starts with digits). If `a` begins with a letter, do the same for leading letters.
+		isDigit := isASCIIDigit(rune(a[vi]))
+
+		var iser func(r rune) bool
+		if isDigit {
+			iser = isASCIIDigit
+		} else {
+			iser = isASCIILetter
+		}
+
+		var ac, bc string
+
+		for _, c := range a[vi:] {
+			if !iser(c) {
+				break
+			}
+
+			ac += string(c)
```

Comment:
Not sure how much this will matter performance wise, but we might want to use stringbuilder here instead.

Reply:
from my testing, using `strings.Builder` makes no difference at all to the performance, and seems to require a couple more lines while looking pretty much the same 🤷 

I've not actually used `strings.Builder` though so maybe I've missed a trick:
```
var sba strings.Builder
for _, c := range a[ai:] {
	if !iser(c) {
		break
	}

	sba.WriteRune(c)
	ai++
}
as := sba.String()

var sbb strings.Builder
for _, c := range b[bi:] {
	if !iser(c) {
		break
	}

	sbb.WriteRune(c)
	bi++
}
bs := sbb.String()
```

<details>
<summary>fwiw, this is my benchmark function</summary>

```
func Benchmark_compareRedHatComponents(b *testing.B) {
	// this is just `runAgainstEcosystemFixture` without the assertion part
	lines := loadEcosystemFixture(b, "redhat-versions-generated.txt")
	versions := make([][]RedHatVersion, 0, len(lines))

	for _, line := range lines {
		pieces := strings.Split(line, " ")

		if len(pieces) != 3 {
			b.Fatalf(`incorrect number of peices in fixture "%s" (got %d)`, line, len(pieces))
		}

		versions = append(versions, []RedHatVersion{
			parseRedHatVersion(pieces[0]),
			parseRedHatVersion(pieces[1]),
		})
	}

	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		for _, v := range versions {
			compareRedHatComponents(v[0].version, v[1].version)
		}
	}
}
```
</details>